tracing/tracing-test.sh

```diff
@@ -72,112 +112,133 @@ stop_jaeger()
 
-	local attempt=0
-	local status=""
+	local traces_url="http://${jaeger_server}:${jaeger_ui_port}/api/traces?service=${service}"
+	curl -s "${traces_url}" 2>/dev/null
+}
 
-	while [ $attempt -lt 10 ]
-	do
-		status=$(curl -s "http://${jaeger_server}:${jaeger_ui_port}/api/traces?service=${service}" 2>/dev/null)
-		local ret=$?
+get_trace_summary()
+{
+	local status="$1"
+	[ -z "$status" ] && die "need jaeger status JSON"
 
-		[ "$ret" -eq 0 ] && [ -n "$status" ] && break
+	echo "${status}" | jq -S '.data[].spans[] | [.spanID, .operationName] | @sh'
+}
 
-		attempt=$((attempt+1))
-		sleep 1
-	done
+get_span_count()
+{
+	local status="$1"
+	[ -z "$status" ] && die "need jaeger status JSON"
 
-	echo "$status"
+	# This could be simplified but creating a variable holding the
+	# summary is useful in debug mode as the summary is displayed.
+	local trace_summary=$(get_trace_summary "$status" || true)
+
+	local count=$(echo "${trace_summary}" | wc -l)
+
+	[ -z "$count" ] && count=0
+
+	echo "$count"
 }
 
-# Look for any "dangling" spans that have not been reported to the Jaeger
-# agent.
-check_missing_spans()
+# Returns status from Jaeger web UI
+get_jaeger_status()
 {
 	local service="$1"
-	local min_spans="$2"
+	local logdir="$2"
 
 	[ -z "$service" ] && die "need jaeger service name"
-	[ -z "$min_spans" ] && die "need minimum trace span count"
+	[ -z "$logdir" ] && die "need logdir"
 
-	local logfile=$(mktemp)
-
-	sudo journalctl -q -o cat -a -t "$service" > "$logfile"
-	sudo chown "$USER:" "$logfile"
+	local attempt=0
+	local max_attempts=10
 
-	# This message needs to be logged by each Kata component, generally when
-	# debug is enabled.
-	local component_prefix="created span"
+	local status=""
+	local span_count=0
 
-	# Message prefix added by Jaeger when LogSpans=true
-	# (see
-	# https://godoc.org/github.com/uber/jaeger-client-go/config#ReporterConfig).
-	local jaeger_reporter_prefix="Reporting span"
+	# Loop until we find some spans
+	while [ "$attempt" -lt "$max_attempts" ]
+	do
+		status=$(get_jaeger_traces "$service" || true)
+		if [ -n "$status" ]; then
+			echo "$status" > "$logdir/${service}-status.json"
+			span_count=$(get_span_count "$status")
+			[ "$span_count" -gt 0 ] && break
+		fi
 
-	local logged_spans
-	if ! logged_spans=$(grep -E -o "${component_prefix} [^ ][^ ]*" "$logfile" | awk '{print $3}')
-	then
-		info "failed to check logged spans"
-		rm -f "$logfile"
-		return
-	fi
+		attempt=$((attempt+1))
+		sleep 1
+	done
 
-	if [ -z "$logged_spans" ]
-	then
-		info "No logged spans to check"
-		rm -f "$logfile"
-		return
-	fi
+	[ -z "$status" ] && die "failed to query Jaeger for status"
+	[ "$span_count" -eq 0 ] && die "failed to find any trace spans"
 
-	local count=0
+	attempt=0
 
-	for span in $logged_spans
+	# Now that we have a valid status, poll until it settles.
+	# Yes, it's horrid but we need to poll to give the Jaeger
+	# web service time to think it seems ;)
+	while [ "$attempt" -lt "$max_attempts" ]
```

Comment:
This can also be `for _ in $(seq $max_attempts)` and attempt removed if `die "span count failed to stabilize after $max_attempts seconds"` is moved out of the loop